app/controllers/v3/isolation_segments_controller.rb

```diff
@@ -86,15 +96,81 @@ def update
+      roles.admin? ||
+      isolation_segment_model.organizations.any? { |org| can_read_from_org?(org.guid) }
+    )
+
+    fetcher = IsolationSegmentOrganizationsFetcher.new(isolation_segment_model)
+    organizations = if roles.admin? || roles.admin_read_only?
+                      fetcher.fetch_all
+                    else
+                      fetcher.fetch_for_organizations(org_guids: readable_org_guids)
+                    end
+
+    render status: :ok, json: Presenters::V3::RelationshipPresenter.new('organizations', organizations)
+  end
+
+  def relationships_spaces
+    isolation_segment_model = IsolationSegmentModel.where(guid: params[:guid]).first
+    resource_not_found!(:isolation_segment) unless isolation_segment_model && (
+      roles.admin? ||
+      isolation_segment_model.spaces.any? { |space| can_read?(space.guid, space.organization.guid) } ||
+      isolation_segment_model.organizations.any? { |org| can_read_from_org?(org.guid) }
+    )
+
+    fetcher = IsolationSegmentSpacesFetcher.new(isolation_segment_model)
+    spaces = if roles.admin? || roles.admin_read_only?
+               fetcher.fetch_all
+             else
+               fetcher.fetch_for_spaces(space_guids: readable_space_guids)
+             end
+
+    render status: :ok, json: Presenters::V3::RelationshipPresenter.new('spaces', spaces)
+  end
 
-  def filter(message, dataset)
-    if message.requested?(:names)
-      dataset = dataset.where(name: message.names)
+  def assign_allowed_organizations
+    unauthorized! unless roles.admin?
+    isolation_segment_model, orgs = organizations_lookup
+
+    isolation_segment_model.db.transaction do
+      isolation_segment_model.lock!
+      orgs.each do |org|
+        isolation_segment_model.add_organization(org)
+      end
     end
-    if message.requested?(:guids)
-      dataset = dataset.where(guid: message.guids)
+
+    render status: :created, json: Presenters::V3::IsolationSegmentPresenter.new(isolation_segment_model)
+  end
+
+  def unassign_allowed_organizations
+    unauthorized! unless roles.admin?
+    isolation_segment_model, orgs = organizations_lookup
+
+    isolation_segment_model.db.transaction do
+      isolation_segment_model.lock!
+      orgs.each do |org|
+        isolation_segment_model.remove_organization(org)
+      end
     end
-    dataset
+
+    head :no_content
+  end
+
+  private
+
+  def organizations_lookup
+    isolation_segment_model = IsolationSegmentModel.where(guid: params[:guid]).first
+    resource_not_found!(:isolation_segment) unless isolation_segment_model
+
+    message = IsolationSegmentAssignOrgMessage.create_from_http_request(params[:body])
+    unprocessable!(message.errors.full_messages) unless message.valid?
+
+    organizations = Organization.where(guid: message.guids).all
+    resources_not_found!("Organization guids: #{message.guids - organizations.map(&:guid)} cannot be found") unless organizations.length == message.guids.length
```

Comment:
https://github.com/cfibmers/cloud_controller_ng/blob/isolation_segments/spec/unit/controllers/v3/isolation_segments_controller_spec.rb#L269-L301  and https://github.com/cfibmers/cloud_controller_ng/blob/isolation_segments/spec/unit/controllers/v3/isolation_segments_controller_spec.rb#L419-L444 hit the resources_not_found! logic.

https://github.com/cfibmers/cloud_controller_ng/blob/isolation_segments/spec/unit/controllers/v3/isolation_segments_controller_spec.rb#L419-L444 (for example) hits the unprocessable! path.